src/charm.py

```diff
@@ -19,6 +24,181 @@ class KratosCharm(CharmBase):
+                    }
+                },
+            },
+            "dsn": f"postgres://{db_info['username']}:{db_info['password']}@{db_info['endpoints']}/postgres",
+            "courier": {
+                "smtp": {
+                    # TODO: dynamic connection uri through charm config
+                    "connection_uri": "smtps://test:test@mailslurper:1025/?skip_ssl_verify=true"
+                }
+            },
+        }
+
+        return yaml.dump(config)
+
+    def _update_layer(self) -> None:
+        """Updates the Pebble configuration layer and kratos config if changed."""
+        # Get current layer
+        current_layer = self._container.get_plan()
+        # Create a new config layer
+        new_layer = self._pebble_layer
+
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self._container.add_layer(self._container_name, new_layer, combine=True)
+            logger.info("Pebble plan updated with new configuration, replanning")
+            self._container.replan()
+
+        # Compare changes in kratos config
+        try:
+            current_config = self._container.pull(self._config_file_path).read()
+        except PathError:
+            self._container.push(self._config_file_path, self._config, make_dirs=True)
+            logger.info("Updated kratos config")
+        else:
+            if current_config != self._config:
+                self._container.push(self._config_file_path, self._config, make_dirs=True)
+                logger.info("Updated kratos config")
+
+        self._container.restart(self._container_name)
+
+    def _get_database_relation_info(self) -> dict:
+        """Get database info from relation data bag."""
+        relation_id = self.database.relations[0].id
+        relation_data = self.database.fetch_relation_data()[relation_id]
+
+        return {
+            "username": relation_data["username"],
+            "password": relation_data["password"],
+            "endpoints": relation_data["endpoints"],
+        }
+
+    def _set_default_identity_schema(self) -> None:
+        """Push identity schema into kratos container."""
+        try:
+            with open("src/identity.default.schema.json", encoding="utf-8") as schema_file:
+                schema = schema_file.read()
+                self._container.push(self._identity_schema_file_path, schema, make_dirs=True)
+            logger.info("Pushed configs to kratos container")
+        except (ProtocolError, PathError) as e:
+            logger.error(str(e))
+            self.unit.status = BlockedStatus(str(e))
+
+    def _update_container(self, event) -> None:
+        """Update configs, pebble layer and run database migration."""
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to Kratos container. Deferring pebble ready event.")
```

Comment:
I think we're deferring an event in general - it could be either pebble ready or database created, since they both use this helper